src/software/ai/hl/stp/tactic/receiver/receiver_fsm.h

```diff
@@ -0,0 +1,296 @@
+
+        // The angle the robot will have to deflect the ball to shoot
+        Angle abs_angle_between_pass_and_shot_vectors;
+        // The percentage of open net the robot would shoot on
+        double net_percent_open;
+        if (best_shot_opt)
+        {
+            Vector robot_to_shot_target =
+                best_shot_opt->getPointToShootAt() - assigned_robot.position();
+            abs_angle_between_pass_and_shot_vectors =
+                (robot_to_ball.orientation() - robot_to_shot_target.orientation())
+                    .clamp()
+                    .abs();
+
+            Angle goal_angle =
+                acuteAngle(world.field().friendlyGoalpostPos(), assigned_robot.position(),
+                           world.field().friendlyGoalpostNeg())
+                    .abs();
+            net_percent_open =
+                best_shot_opt->getOpenAngle().toDegrees() / goal_angle.toDegrees();
+        }
+
+        // If we have a shot with a sufficiently large enough opening, and the
+        // deflection angle that is reasonable, we should one-touch kick the ball
+        // towards the enemy net
+        if (best_shot_opt && net_percent_open > MIN_SHOT_NET_PERCENT_OPEN &&
+            abs_angle_between_pass_and_shot_vectors < MAX_DEFLECTION_FOR_ONE_TOUCH_SHOT)
+        {
+            return best_shot_opt;
+        }
+        return std::nullopt;
+    }
+
+
+    auto operator()()
+    {
+        using namespace boost::sml;
+
+        const auto undecided_s = state<UndecidedState>;
+        const auto receive_s   = state<ReceiveAndDribbleState>;
+        const auto onetouch_s  = state<OneTouchShotState>;
+        const auto update_e    = event<Update>;
+
+        /**
+         * Checks if a one touch shot is possible
+         *
+         * @param event ReceiverFSM::Update event
+         * @return true if one-touch possible
+         */
+        const auto onetouch_possible = [](auto event) {
+            return findFeasibleShot(event.common.world, event.common.robot) !=
+                   std::nullopt;
+        };
+
+        /**
+         * If we have a shot on net, then update the receiver fsm
+         * to setup for a one-touch shot.
+         *
+         * NOTE: This must be used with the onetouch_possible guard,
+         * which checks for one-touch feasibility.
+         *
+         * @param event ReceiverFSM::Update event
+         */
+        const auto update_onetouch = [this](auto event) {
+            auto best_shot = findFeasibleShot(event.common.world, event.common.robot);
+            auto one_touch = getOneTimeShotPositionAndOrientation(
+                event.common.robot, event.common.world.ball(),
+                best_shot->getPointToShootAt());
+
+            if (best_shot)
+            {
+                if (event.control_params.pass)
+                {
+                    event.common.set_intent(std::make_unique<MoveIntent>(
```

Comment:
why would we use an intent here as opposed to KickFSM?

Reply:
its because we need 2 states, 1 for one-touch (which would need a MoveFSM) and another for receive and dribble which would be another MoveFSM

since we can't have two of the same FSMs used twice in the same FSM, we have to use the `MoveIntent`

The KickFSM also is slower, it has to get behind ball, and then kick. Orientating in the right position and turning on auto-kick is definitely faster :tm: